lib/Lower/PFTBuilder.cpp

```diff
@@ -237,45 +239,514 @@ class PFTBuilder {
+                  labelSet.insert(label);
+                  assignSymbolLabelMap->try_emplace(sym, labelSet);
+                } else {
+                  iter->second.insert(label);
+                }
+              },
+              [&](const parser::AssignedGotoStmt *) {
+                // Specific control successors are not in general known.
+                // Compensate by directly marking the successor new block.
+                markSuccessorNewBlock(eval);
+              },
+
+              // Construct statements
+              [&](const parser::AssociateStmt *s) {
+                insertConstructName(&*s, parentConstruct);
+              },
+              [&](const parser::BlockStmt *s) {
+                insertConstructName(&*s, parentConstruct);
+              },
+              [&](const parser::SelectCaseStmt *s) {
+                insertConstructName(&*s, parentConstruct);
+                eval.lexicalSuccessor->isNewBlock = true;
+              },
+              [&](const parser::CaseStmt *) { eval.isNewBlock = true; },
+              [&](const parser::ChangeTeamStmt *s) {
+                insertConstructName(&*s, parentConstruct);
+              },
+              [&](const parser::CriticalStmt *s) {
+                insertConstructName(&*s, parentConstruct);
+              },
+              [&](const parser::NonLabelDoStmt *s) {
+                insertConstructName(&*s, parentConstruct);
+                doConstructStack.push_back(parentConstruct);
+                auto &control{
+                    std::get<std::optional<parser::LoopControl>>(s->t)};
+                // eval.block is the loop preheader block, which will be set
+                // elsewhere if the NonLabelDoStmt is itself a target.
+                // eval.localBlocks[0] is the loop header block.
+                eval.localBlocks.emplace_back(nullptr);
+                if (!control.has_value()) {
+                  eval.isUnstructured = true; // infinite loop
+                  return;
+                }
+                eval.lexicalSuccessor->isNewBlock = true;
+                eval.controlSuccessor = &evaluationList.back();
+                if (std::holds_alternative<parser::ScalarLogicalExpr>(
+                        control->u)) {
+                  eval.isUnstructured = true; // while loop
+                }
+                // Defer additional processing for a concurrent loop to the
+                // EndDoStmt, when it is known if the loop is structured or not.
+              },
+              [&](const parser::EndDoStmt *) {
+                pft::Evaluation &doEval{evaluationList.front()};
+                eval.controlSuccessor = &doEval;
+                doConstructStack.pop_back();
+                if (parentConstruct->lowerAsStructured()) {
+                  return;
+                }
+                parentConstruct->constructExit->isNewBlock = true;
+                const auto &doStmt{doEval.getIf<parser::NonLabelDoStmt>()};
```

Comment:
I'm not sure why this seems misleading.  In case that is because this code is mostly using NonLabelDoStmt information in an EndDoStmt lambda, I've added another comment about that.

Reply:
`const auto&` is a bit misleading because usually people do not expect the resulting type to be a pointer, but here `getIf` returns a pointer. I was just suggesting `const auto*` instead of `const auto&`. No big deal though.